### ISSUE TITLE
Problem with cancelling touches in CCScrollLayer

### ISSUE THREAD

**Issue by mrwilczarz** [cocos2d/cocos2d-iphone-extensions#38]
Hi, 

Please reconsider calling the folllowing in ccTouchMoved method of CCScrollLayer.m:

``` objectivec
[[CCTouchDispatcher sharedDispatcher] touchesCancelled: [NSSet setWithObject: touch] withEvent:event];
```

The problem is that if I provide implementation of ccTouchCancelled, it will be called during the first ccTouchMoved invocation, and in my opinion it should only be called when the touch is interrupted by external event (system message, phone call etc.)

Consider the following example: the scroll layer holds the first scrolling touch to avoid multiple touch issues (see http://www.cocos2d-iphone.org/forum/topic/17118) With this approach I really need following method, to release the touch when incoming call appears:

``` objectivec
-(void)ccTouchCancelled:(UITouch *)touch withEvent:(UIEvent *)event {
    if( scrollTouch_ == touch ) {
        scrollTouch_ = nil;
        [self selectPage: currentScreen_];
    }
}
```

However, when present, ccTouchCancelled will be called just after scroll starts and therefore, no scrolling animation occurs.

Thanks!

**Comment by psineur** [comment]
Now i got it, thanks!

Since CCTouchDispatcher doesn't queue touches, it's possible to fix this simply by setting a flag before & unsetting it after https://github.com/cocos2d/cocos2d-iphone-extensions/blob/develop/Extensions/CCScrollLayer/CCScrollLayer.m#L311

This flag will be used for test before reciting on ccTouchesCancelled. Gonna fix it for 0.1.2
Also! In 0.1.2 there will be delegate for start/stop scrolling and ability to disable stealing touch. So it will be possible to work around scrolling above active target delegates without invoking touchesCancelled method.

**Comment by psineur** [comment]
BTW,I will fix this in few days so if you want to try to play with GitHub & pull requests - try this and send me a pull request if it works.

``` objectivec
- (void) cancelAndStoleTouch:(UITouch *)touch withEvent:(UIEvent *)event
{


    // Throw Cancel message for everybody in TouchDispatcher and do not react on this.
        stealingTouchInProgress_ = YES;
    [[CCTouchDispatcher sharedDispatcher] touchesCancelled: [NSSet setWithObject: touch] withEvent:event];
        stealingTouchInProgress_ = NO;

    //< after doing this touch is already removed from all targeted handlers

    // Squirrel away the touch
    [self claimTouch: touch];
}

-(void)ccTouchCancelled:(UITouch *)touch withEvent:(UIEvent *)event 
{
    // Do not cancell touch, if this method is called from cancelAndStoleTouch:
    if (stealingTouchInProgress_)
          return;

    if( scrollTouch_ == touch ) {
        scrollTouch_ = nil;
        [self selectPage: currentScreen_];
    }
}
```

```

How to make pull request:
1. Fork extensions on GitHub
2. Create new branch "issue-38-fix" from latest state of develop branch   
     git checkout -t origin/develop   
     git checkout -b issue-38-fix

3. Make changes and commit them. Use Fixes #38 in last commit name to mark this issue as fixed.
4. push it to YOUR repo on the GitHub:   
    git push -u origin issue-38-fix

5. Create a pull request from your issue-38-fix branch to cocos2d/cocos2d-iphone-extensions develop branch
(make sure you checked develop instead of master)

If you'll do everything correct - it will be possible for me to merge your pull request in just one click
 & issue will be closed by you. 

P.S.
Make sure you have configured git for github right ;)
```

**Comment by psineur** [comment]
TODO: make sure to have ability to cancel stealing touches on iOS.
